[DSIP-55][Master] Separate the waiting dispatched task into different queue by worker group

det101 opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler, we are happy that you want to help us improve DolphinScheduler! -->

## Purpose of the pull request
Simplify the GlobalTaskDispatchWaitingQueue to only handle sorting based on time.
Introduce a new WorkerGroup queue that sorts tasks based on priority and time, with priority being higher than time. Additionally, it is responsible for retrying failed tasks with a delay.

close #16260

## Brief change log



This change added tests and can be verified as follows:

- *Manually verified the change by testing locally.*



## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contains incompatible change, you should also add it to `docs/docs/en/guide/upgrade/incompatible.md`

Commit:
fix #16260
Simplify the GlobalTaskDispatchWaitingQueue to only handle sorting based on time.
Introduce a new WorkerGroup queue that sorts tasks based on priority and time, with priority being higher than time. Additionally, it is responsible for retrying failed tasks with a delay.

Commit:
fix code style

Commit:
fix spotless

Commit:
fix unit test

Commit:
fix spotless

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/WorkflowEngine.java:
```diff
@@ -42,6 +43,9 @@ public class WorkflowEngine implements AutoCloseable {
     private LogicTaskEngineDelegator logicTaskEngineDelegator;
 
+    @Autowired
+    private WorkerGroupQueueLooper workerGroupQueueLooper;
```
```suggestion
    private WorkerGroupTaskDispatchWaitingQueueLooper workerGroupTaskDispatchWatingQueueLooper;
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueue.java:
```diff
@@ -31,18 +32,19 @@
-    private final PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>> priorityDelayQueue =
-            new PriorityDelayQueue<>();
+
+    private final PriorityBlockingQueue<DelayEntry<ITaskExecutionRunnable>> priorityQueue =
```
You need to use `DelayQueue` here? Otherwise how to make sure the item will be poll out by a delay time.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueue.java:
```diff
@@ -57,19 +59,21 @@ public synchronized void dispatchTaskExecuteRunnable(ITaskExecutionRunnable ITas
     @SneakyThrows
-    public ITaskExecutionRunnable takeTaskExecuteRunnable() {
-        ITaskExecutionRunnable taskExecutionRunnable = priorityDelayQueue.take().getData();
+    public DelayEntry<ITaskExecutionRunnable> takeTaskExecuteRunnable() {
```
```suggestion
    public ITaskExecutionRunnable takeTaskExecuteRunnable() {
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueueLooper.java:
```diff
@@ -38,8 +38,7 @@ public class GlobalTaskDispatchWaitingQueueLooper extends BaseDaemonThread imple
     @Autowired
-    private ITaskExecutorClient taskExecutorClient;
-
+    private WorkerGroupQueueMap workerGroupQueueMap;
```
```suggestion
    private WorkerGroupTaskDispatchWaitingQueues workerGroupTaskDispatchWaitingQueues;
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueueLooper.java:
```diff
@@ -65,25 +64,10 @@ public void run() {
-                    waitingTimeMills);
-            log.error("Dispatch Task: {} failed will retry after: {}/ms", taskInstance.getName(), waitingTimeMills, e);
-        }
+        DelayEntry<ITaskExecutionRunnable> delayEntry = globalTaskDispatchWaitingQueue.takeTaskExecuteRunnable();
```
Need to add log here.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupQueueLooper.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Override
+    public void run() {
+        while (RUNNING_FLAG.get()) {
+            doDispatch();
+        }
```
Will make CPU 100%

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/queue/DelayEntry.java:
```diff
@@ -54,14 +54,6 @@ public long getDelay(@NotNull TimeUnit unit) {
     @Override
     public int compareTo(@NotNull Delayed o) {
         DelayEntry<V> other = (DelayEntry<V>) o;
-        int delayTimeMillsCompareResult = Long.compare(delayTimeMills, other.delayTimeMills);
-        if (delayTimeMillsCompareResult != 0) {
-            return delayTimeMillsCompareResult;
-        }
-
-        if (data == null || other.data == null) {
-            return 0;
-        }
```
Don't do this kind of change, this class is `DelayEntry` if you remove the compare logic this class cannot work

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/queue/PriorityDelayEntry.java:
```diff
@@ -17,25 +17,24 @@
+import org.jetbrains.annotations.NotNull;
 
-public class PriorityDelayQueue<V extends DelayEntry> {
+public class PriorityDelayEntry<V extends Comparable<V>> extends DelayEntry {
```
Don't change this class.

det101 replied:
Already restored

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/queue/WorkerGroupQueueMap.java:
```diff
@@ -0,0 +1,58 @@
+@Component
+public class WorkerGroupQueueMap {
+
+    private final Map<String, DelayQueue<PriorityDelayEntry<ITaskExecutionRunnable>>> queueMap = new HashMap<>();
```
```suggestion
    private final Map<String, DelayQueue<PriorityDelayEntry<ITaskExecutionRunnable>>> queueMap = new ConcurrentHashMap<>();
```

det101 replied:
The modification has been completed

ruanwenjun commented on dolphinscheduler-ui/src/router/modules/ui-setting.ts:
```diff
@@ -30,11 +30,11 @@ export default {
     {
       path: '',
-      name: 'ui-setting',
+      name: 'ui-setting2',
```
Please use another PR to do this change.

det101 replied:
Removed modifications

Commit:
WorkerGroupTaskDispatchManager manages threads and queues.
DispatchWorker is responsible for the specific logic of dispatching tasks.
WorkerGroupTaskDispatchWaitingQueueLooper is the thread that handles task dispatching.
ComparableEntry is used only for priority comparison.

Commit:
WorkerGroupTaskDispatchManager manages threads and queues.
DispatchWorker is responsible for the specific logic of dispatching tasks.
WorkerGroupTaskDispatchWaitingQueueLooper is the thread that handles task dispatching.
ComparableEntry is used only for priority comparison.

Commit:
rollback master logback.xml

Commit:
fix style

Commit:
close workerGroupTaskDispatchManager

Commit:
close workerGroupTaskDispatchManager

Commit:
fix spotless

Commit:
add ThreadCreatingAndDestroyingWorkerGroupListener

Commit:
add unit test

Commit:
add unit test

Commit:
add unit test

Commit:
optimized code

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/queue/ComparableEntry.java:
```diff
@@ -0,0 +1,68 @@
+import org.jetbrains.annotations.NotNull;
+
+@Getter
+public class ComparableEntry implements Comparable<ComparableEntry> {
```
```suggestion
public class TimeBasedTaskExecutionRunnableComparableEntry implements Comparable<ComparableEntry> {
```

det101 replied:
This suggestion clarifies the ability to compare and contrast, and the changes have been made.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/queue/ComparableEntry.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Override
+    public int compareTo(@NotNull ComparableEntry other) {
+        int priortyCompareResult = data.compareTo(other.data);
+        if (priortyCompareResult != 0) {
+            return priortyCompareResult;
+        }
+        return Long.compare(delayTimeMills, other.delayTimeMills);
+    }
```
```suggestion
    public int compareTo(@NotNull ComparableEntry other) {
        int delayTimeCompareResult = Long.compare(delayTimeMills, other.delayTimeMills);
        if (delayTimeCompareResult != 0) {
            return delayTimeCompareResult;
        }
        return data.compareTo(other.data);
    }
```

As discussed at https://github.com/apache/dolphinscheduler/issues/16260#issuecomment-2647236911, this comparation will only compare the delay time.

det101 replied:
Thank you for pointing out the error. It has been changed to a logic consistent with the design

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/queue/ComparableEntry.java:
```diff
@@ -0,0 +1,68 @@
+@Getter
+public class ComparableEntry implements Comparable<ComparableEntry> {
+
+    // Pass it to the workerGroup queue without participating in the comparison
```
```suggestion
    // Pass it to the workerGroup queue without participating in the comparison
```

det101 replied:
removed

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatchWaitingQueueLooper.java:
```diff
@@ -0,0 +1,78 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class WorkerGroupTaskDispatchWaitingQueueLooper extends BaseDaemonThread implements AutoCloseable {
```
Remove this class, directly use DispatchWorker

det101 replied:
removed

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatchManager.java:
```diff
@@ -0,0 +1,117 @@
+
+@Component
+@Slf4j
+public class WorkerGroupTaskDispatchManager implements AutoCloseable {
```
```suggestion
public class WorkerGroupTaskDispatcherManager implements AutoCloseable {
```

det101 replied:
Modified

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatchManager.java:
```diff
@@ -0,0 +1,117 @@
+    private ITaskExecutorClient taskExecutorClient;
+
+    @Getter
+    private final ConcurrentHashMap<String, WorkerGroupTaskDispatchWaitingQueueLooper> workerGroupTaskDispatchWaitingQueueLooperMap;
```
```suggestion
    private final ConcurrentHashMap<String, WorkerGroupTaskDispatcher> workerGroupTaskDispatchWaitingQueueLooperMap;
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatchManager.java:
```diff
@@ -0,0 +1,117 @@
+
+    @Getter
+    private final ConcurrentHashMap<String, WorkerGroupTaskDispatchWaitingQueueLooper> workerGroupTaskDispatchWaitingQueueLooperMap;
+    @Getter
+    private final ConcurrentHashMap<String, PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>>> workerGroupPriorityDelayQueueMap;
+
```
```suggestion
    @Getter
    private final ConcurrentHashMap<String, PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>>> workerGroupPriorityDelayQueueMap;

```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatchManager.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    /**
+     * Adds a task to the specified worker group queue and starts or wakes up the corresponding processing loop.
```
```suggestion
     * Adds a task to the specified worker group queue and starts or wakes up the corresponding processing loop.
```

There is no wake up logic.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatchManager.java:
```diff
@@ -0,0 +1,117 @@
+     *
+     * @param workerGroup the identifier for the worker group
+     */
+    public synchronized void stopWorkerGroup(String workerGroup) throws Exception {
+        WorkerGroupTaskDispatchWaitingQueueLooper looper =
+                workerGroupTaskDispatchWaitingQueueLooperMap.remove(workerGroup);
+        PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>> workerGroupQueue =
+                workerGroupPriorityDelayQueueMap.get(workerGroup);
+        if (workerGroupQueue != null) {
+            workerGroupQueue.clear();
+        }
+        if (looper != null) {
+            looper.close();
+        }
+    }
```
There will exist a problem, you cannot clear the task, the running task which put into the queue should continue running.

det101 replied:
This method is a workergroup deletion logic, which may not be able to handle even if it is placed back in the global queue

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueue.java:
```diff
@@ -57,19 +59,22 @@ public synchronized void dispatchTaskExecuteRunnable(ITaskExecutionRunnable ITas
     @SneakyThrows
-    public ITaskExecutionRunnable takeTaskExecuteRunnable() {
-        ITaskExecutionRunnable taskExecutionRunnable = priorityDelayQueue.take().getData();
+    public ComparableEntry takeTaskExecuteRunnable() {
```
```suggestion
    public ITaskExecutionRunnable takeTaskExecuteRunnable() {
```

det101 replied:
I have already modified

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/DispatchWorker.java:
```diff
@@ -0,0 +1,63 @@
+
+    private final ITaskExecutorClient taskExecutorClient;
+
+    private final PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>> workerGroupQueue;
```
Please add a `PriorityBasedTaskExecutionRunnableDelayEntry`, this entry should only compare priority rather than delay time.

det101 replied:
I have a question. If only priority is compared, then tasks that fail and need to be retried will not have any delay before being retried.

det101 replied:
![image](https://github.com/user-attachments/assets/4cd4f203-2701-4a22-abaf-5096796f9bd9)

det101 replied:
// todo: The current queue is flawed. When a high-priority task fails,
    // it will be delayed and will not return to the first or second position.
    // Tasks with the same priority will preempt its position.
    // If it needs to be placed at the front of the queue, the queue needs to be re-implemented.
    private final PriorityDelayQueue<PriorityAndDelayBasedTaskEntry> workerGroupQueue;

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/cluster/ThreadCreatingAndDestroyingWorkerGroupListener.java:
```diff
@@ -0,0 +1,72 @@
+
+@Component
+@Slf4j
+public class ThreadCreatingAndDestroyingWorkerGroupListener implements WorkerGroupChangeNotifier.WorkerGroupListener {
```
Remove this class, `WorkerGroupTaskDispatchManager` should implements with `WorkerGroupChangeNotifier.WorkerGroupListener`;

det101 replied:
removed

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/cluster/ClusterManager.java:
```diff
@@ -48,6 +49,10 @@ public class ClusterManager {
     private RegistryClient registryClient;
 
+    @Autowired
+    @Lazy
```
Please remove `@Lazy`, this is not a good practice, we should avoid this.

det101 replied:
remove @Lazy

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueue.java:
```diff
@@ -31,24 +32,25 @@
-    private final PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>> priorityDelayQueue =
-            new PriorityDelayQueue<>();
+
+    private final PriorityBlockingQueue<TimeBasedTaskExecutionRunnableComparableEntry> priorityQueue =
```
```suggestion
    private final PriorityBlockingQueue<TimeBasedTaskExecutionRunnableComparableEntry> delayQueue =
```

det101 replied:
Okay, this should be the delay queue. I will change the name

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueue.java:
```diff
@@ -57,19 +59,22 @@ public synchronized void dispatchTaskExecuteRunnable(ITaskExecutionRunnable ITas
     @SneakyThrows
-    public ITaskExecutionRunnable takeTaskExecuteRunnable() {
-        ITaskExecutionRunnable taskExecutionRunnable = priorityDelayQueue.take().getData();
+    public TimeBasedTaskExecutionRunnableComparableEntry takeTaskExecuteRunnable() {
```
Why not return `ITaskExecutionRunnable`?  Why expose `TimeBasedTaskExecutionRunnableComparableEntry` to outside?

det101 replied:
The previous global queue data retrieval solution was incorrect, but it has now been modified.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/cluster/ClusterManager.java:
```diff
@@ -48,9 +50,13 @@ public class ClusterManager {
     public ClusterManager() {
         this.masterClusters = new MasterClusters();
         this.workerClusters = new WorkerClusters();
+        workerGroupTaskDispatchManager = new WorkerGroupTaskDispatchManager();
```
Can this work? `WorkerGroupTaskDispatchManager` is a spring bean, did you test this?

det101 replied:
Sorry, I missed testing it. I have already modified

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/DispatchWorker.java:
```diff
@@ -0,0 +1,107 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class DispatchWorker extends BaseDaemonThread implements AutoCloseable {
```
```suggestion
public class WorkerGroupTaskDispatcher extends BaseDaemonThread implements AutoCloseable {
```

det101 replied:
The modification has been completed

Commit:
fix lazy

Commit:
add todo

Commit:
fix spotless:

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/cluster/ClusterManager.java:
```diff
@@ -48,12 +49,15 @@ public class ClusterManager {
     }
 
-    public void start() {
+    public void start(WorkerGroupTaskDispatchManager workerGroupTaskDispatchManager) {
+        this.workerGroupTaskDispatchManager = workerGroupTaskDispatchManager;
```
You should add a method to register the `WorkerGroupListener` rather than change the start method, this is too strange.

det101 replied:
i will add registerWorkerGroupListener function

Commit:
change WorkerGroupTaskDispatchManager -> WorkerGroupTaskDispatcherManager

Commit:
fix sonar

Commit:
optimized code

Commit:
optimized code

Commit:
global queue is delayQueue

Commit:
PriorityDelayQueue move to WorkerGroupTaskDispatcher,Simplify the WorkerGroupTaskDispatcherManager logic

Commit:
Remove unnecessary resource cleanup

Commit:
less unit test time

Commit:
fix jdk8 unit test

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/cluster/ClusterManager.java:
```diff
@@ -48,6 +49,8 @@ public class ClusterManager {
     @Autowired
     private RegistryClient registryClient;
 
+    private WorkerGroupTaskDispatcherManager workerGroupTaskDispatcherManager;
```
```suggestion
```

ruanwenjun replied:
Use a list here or we can directly remove this line.

det101 replied:
Thank you very much for your feedback! I will fixe this。

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/cluster/ClusterManager.java:
```diff
@@ -59,6 +62,10 @@ public void start() {
     }
 
+    public void registerWorkerGroupListener(WorkerGroupTaskDispatcherManager workerGroupTaskDispatcherManager) {
+        this.workerGroupTaskDispatcherManager = workerGroupTaskDispatcherManager;
```
```suggestion
        this.workerGroupChangeNotifier.subscribeWorkerGroupsChange(workerGroupTaskDispatcherManager);
```

det101 replied:
Thank you very much for your feedback! I will fixe this。

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueue.java:
```diff
@@ -31,24 +32,25 @@
-    private final PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>> priorityDelayQueue =
-            new PriorityDelayQueue<>();
+
+    private final PriorityBlockingQueue<TimeBasedTaskExecutionRunnableComparableEntry> delayQueue =
```
We should use DelayQueue here.

det101 replied:
Thank you very much for your feedback! I will fixe this。

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/queue/PriorityAndDelayBasedTaskEntry.java:
```diff
@@ -0,0 +1,50 @@
+
+import org.jetbrains.annotations.NotNull;
+
+public class PriorityAndDelayBasedTaskEntry<V extends Comparable<V>> extends DelayEntry<V> {
```
```suggestion
public class PriorityDelayTaskEntry<V extends Comparable<V>> extends DelayEntry<V> {
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcher.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Override
+    public void close() throws Exception {
```
```suggestion
    public void markDispatcherDeleting() throws Exception {
```

ruanwenjun replied:
You need to do concurrency control in this class.

det101 replied:
Since the close method is overridden, I plan to move the logic to markDispatcherDeleting. Regarding concurrency, I will add synchronized to the close method.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcher.java:
```diff
@@ -0,0 +1,125 @@
+        }
+    }
+
+    public void dispatch() {
```
```suggestion
    private void dispatch() {
```

ruanwenjun replied:
This method should not exposed to outside.

det101 replied:
thanks，i will fix this

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,185 @@
+    @Getter
+    private final ConcurrentHashMap<String, WorkerGroupTaskDispatcher> dispatchWorkerMap;
+
+    private final ScheduledExecutorService scheduler;
```
You cannot control this threadpool, so don't maintain this.
```suggestion
```

det101 replied:
ok，I will use MasterThreadFactory.getDefaultSchedulerThreadExecutor() instead.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/DispatchWorkerStatus.java:
```diff
@@ -0,0 +1,24 @@
+package org.apache.dolphinscheduler.server.master.runner;
+
+public enum DispatchWorkerStatus {
+    DEFAULT,
+    DELETE_SUCCESS,
+    DELETING
```
```suggestion
    STARTED,
    CLOSING,
    CLOSED,
```

det101 replied:
Thank you very much for your feedback! I will fixe this。

ruanwenjun reviewed (changes requested):
miss

Commit:
optimized code

Commit:
fix sonar

Commit:
override equals

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueue.java:
```diff
@@ -31,24 +32,25 @@
      */
-    public synchronized void dispatchTaskExecuteRunnable(ITaskExecutionRunnable ITaskExecutionRunnable) {
-        dispatchTaskExecuteRunnableWithDelay(ITaskExecutionRunnable, 0);
+    public synchronized void dispatchTaskExecuteRunnable(ITaskExecutionRunnable iTaskExecutionRunnable) {
+        dispatchTaskExecuteRunnableWithDelay(iTaskExecutionRunnable, 0);
```
Why the delay time is 0 here, if so the `delayQueue` is meaningless.

det101 replied:
Global queue sorted by time, this method has been removed

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,182 @@
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher != null) {
+            workerGroupTaskDispatcher.add(taskExecutionRunnable, delayTimeMills);
+            log.info("queue size {}", workerGroupTaskDispatcher.size());
```
```suggestion
            
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,182 @@
+        } else {
+            log.error("workerGroupTaskDispatcher {} not found", workerGroup);
+        }
+        return false;
```
If the WorkerGroupTaskDispatcher doens't exist, then the task lifecycle will break? I don't see anywhere deal with the result.

det101 replied:
cannot find the workergroup, task will kill.

SbloodyS replied:
> cannot find the workergroup, task will kill.

This conflicts with the existing waiting rules. We should be consistent with the existing logic.

det101 replied:
> > cannot find the workergroup, task will kill.
> 
> This conflicts with the existing waiting rules. We should be consistent with the existing logic.


If the name of the worker group is changed, is there a situation where the original logical task will never be completed?

det101 replied:
> > cannot find the workergroup, task will kill.
> 
> This conflicts with the existing waiting rules. We should be consistent with the existing logic.
Already changed to existing logic

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcher.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Override
+    public synchronized void start() {
+        if (!RUNNING_FLAG.compareAndSet(false, true)) {
+            log.error("The {} already started, will not start again", this.getName());
+            return;
+        }
+        log.info("{} starting...", this.getName());
+        super.start();
+        log.info("{} started...", this.getName());
+    }
```
If a dispatcher closed and then start again, can it start twice? this is a thread!

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,182 @@
+        }
+    }
+
+    private void checkDeleteDispatchWorkerComplete() {
+        for (Map.Entry<String, WorkerGroupTaskDispatcher> entry : dispatchWorkerMap.entrySet()) {
+            String workerGroup = entry.getKey();
+            WorkerGroupTaskDispatcher workerGroupTaskDispatcher = entry.getValue();
+            switch (workerGroupTaskDispatcher.getStatus()) {
+                case CLOSING:
+                    try (WorkerGroupTaskDispatcher ignored = workerGroupTaskDispatcher) {
+                        log.info("try to delete worker group {}", workerGroup);
+                    } catch (Exception e) {
+                        log.error("stop worker group error", e);
+                    }
+                    break;
+                case CLOSED:
+                    try (WorkerGroupTaskDispatcher ignored = dispatchWorkerMap.remove(workerGroup)) {
+                        log.info("success remove worker group {}", workerGroup);
+                    } catch (Exception e) {
+                        log.error("stop worker group error", e);
+                    }
+                    break;
+                default:
+                    log.debug("worker group {} status {}", workerGroup, workerGroupTaskDispatcher.getStatus());
+                    break;
+            }
+        }
+    }
```
Once a dispatcher's status is closed, then it might still exist task in its queue, I don't see any control about this.

ruanwenjun reviewed (changes requested):
You'd better think carefully about the logic here, this is likely to cause serious problems

Commit:
Optimize concurrency issues
The logic for the non-existence of workergroup needs to be supplemented.

Commit:
\cannot find the workergroup, task will kill.

Commit:
fix unit test

Commit:
fix unit test

Commit:
spotless

Commit:
retry e2e

Commit:
jdk 11 unit test

Commit:
Maintain the original logic: if the worker group cannot be found, place it back into the queue and wait.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueueLooper.java:
```diff
@@ -65,24 +66,20 @@ public void run() {
+                taskExecutionRunnable.getTaskInstance().getWorkerGroup(),
+                taskExecutionRunnable, 0);
+        if (!addTaskSuccess) {
+            log.warn("worker group is deleting or deleted, taskInstance: {}", taskExecutionRunnable.getTaskInstance());
```
```suggestion
            log.warn("Dispatch TaskInstance: {} WorkerGrouTaskDispatcher: {} failed", taskExecutionRunnable.getTaskInstance().getName(), taskExecutionRunnable.getTaskInstance().getWorkerGroup());
```

det101 replied:
Thank you very much for your suggestions! I completely agree with your changes and they do make the code clearer. I will make the changes according to your suggestions.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueueLooper.java:
```diff
@@ -65,24 +66,20 @@ public void run() {
-        final ITaskExecutionRunnable taskExecutionRunnable = globalTaskDispatchWaitingQueue.takeTaskExecuteRunnable();
-        final TaskInstance taskInstance = taskExecutionRunnable.getTaskInstance();
-        try {
-            final TaskExecutionStatus status = taskInstance.getState();
-            if (status != TaskExecutionStatus.SUBMITTED_SUCCESS && status != TaskExecutionStatus.DELAY_EXECUTION) {
```
Why you remove this? Are you clear on what this code does? After your removed, the task instance cannot be removed after killed.

det101 replied:
Thank you for your concern. I have moved the status check to the WorkerGroupTaskDispatcher, which should prevent the task instance from not being removable after being killed. However, I understand that moving it to the global queue for an early check is also a viable option. I appreciate your feedback and will keep this in mind for further improvements.

det101 replied:
Thank you for bringing this up. I have moved the status check to the WorkerGroupTaskDispatcher. I believe this change should ensure that the task instance can be removed after being killed.   Placing the status check in the global queue could indeed be a better approach. I’m happy to explore that option further if you think it would improve the process.

det101 replied:
> Thank you for bringing this up. I have moved the status check to the WorkerGroupTaskDispatcher. I believe this change should ensure that the task instance can be removed after being killed. Placing the status check in the global queue could indeed be a better approach. I’m happy to explore that option further if you think it would improve the process.

 After some consideration, I have decided to add status checks in the global queue as well.

det101 replied:
Remove the close logic for WorkerGroupTaskDispatcher. Since this scenario is rare and the logic is simplified for now, it will be left as a TODO.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueueLooper.java:
```diff
@@ -91,6 +88,7 @@ public void close() throws Exception {
         if (RUNNING_FLAG.compareAndSet(true, false)) {
             log.info("GlobalTaskDispatchWaitingQueueLooper stopping...");
             log.info("GlobalTaskDispatchWaitingQueueLooper stopped...");
+            workerGroupTaskDispatcherManager.close();
```
Should before `log.info("GlobalTaskDispatchWaitingQueueLooper stopped...");`

det101 replied:
My mistake led to the meaningless log. I will make the changes as suggested.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcher.java:
```diff
@@ -0,0 +1,146 @@
+            workerGroupQueue.add(new PriorityAndDelayBasedTaskEntry(delayTimeMills, taskExecutionRunnable));
+            return true;
+        } else {
+            log.warn("The {} status is {}, task can not add Queue, it will fail", this.getName(), status.get());
```
```suggestion
            log.warn("The WorkerGroupTaskDispatcher: {} status is {}, cannot receive task: {}", this.getName(), status.get(), taskExecutionRunnable.getTaskInstance().getName());
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcher.java:
```diff
@@ -0,0 +1,146 @@
+        }
+    }
+
+    public void markDispatcherStart() {
+        if (status.compareAndSet(DispatchWorkerStatus.CLOSING, DispatchWorkerStatus.STARTED)) {
+            log.info("The {} markDispatcherStart...", this.getName());
+        } else {
+            log.warn("The {} status is {}, will not markDispatcherStart", this.getName(), status.get());
+        }
+    }
```
If the thread has already died, this method cannot restart the dispatcher again.

det101 replied:
I see. In my design, the dispatcher can still be restarted if it is in the closing state. I just want to make sure I understand correctly—are you saying that once it’s marked as closed, it should no longer support restarting?

det101 replied:
> I see. In my design, the dispatcher can still be restarted if it is in the closing state. I just want to make sure I understand correctly—are you saying that once it’s marked as closed, it should no longer support restarting?

@ruanwenjun

det101 replied:
Remove the markstart logic for WorkerGroupTaskDispatcher. Since this scenario is rare and the logic is simplified for now

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,174 @@
+            dispatchWorkerMap.put(workerGroup, workerGroupTaskDispatcher);
+            workerGroupTaskDispatcher.start();
+        } else {
+            workerGroupTaskDispatcher.markDispatcherStart();
```
There will exist concurrency problem, if the dispatcher just close the loop.

det101 replied:
Since markDispatcherClosing is not a frequently called method, I will add the synchronized keyword to ensure thread safety and address concurrency issues.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,174 @@
+     */
+    @Override
+    public void close() throws Exception {
+        log.info("WorkerGroupTaskDispatcherManager start close");
+        for (Map.Entry<String, WorkerGroupTaskDispatcher> entry : dispatchWorkerMap.entrySet()) {
+            try {
+                entry.getValue().markDispatcherClosing();
+            } catch (Exception e) {
+                log.error("stop worker group error", e);
+            }
+        }
+        log.info("WorkerGroupTaskDispatcherManager closed");
```
In fact you should closed the dispatcher here not just closing.

det101 replied:
change to 
```
 entry.getValue().close();
```
WorkerGroupTaskDispatcher will be left as a TODO.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Override
+    public void onWorkerGroupChange(List<WorkerGroup> workerGroups) {
+        this.checkAndRemoveClosedDispatchWorker();
+        String workerGroupsString = workerGroups.stream()
+                .map(WorkerGroup::getName)
+                .collect(Collectors.joining(", "));
+        log.info("Worker groups: {}", workerGroupsString);
+    }
```
The logic is too strange, and the log is meaningless, no one can unstandard what the log means if he doesn't familiar with the code.

det101 replied:
There is no need to handle the change events here; just log the records. I will change it to log.info("on change worker groups: {}", workerGroups);

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,174 @@
+            try {
+                this.deleteWorkerGroup(workerGroup.getName());
+            } catch (Exception e) {
+                log.error("stop worker group error", e);
```
```suggestion
                log.error("Delete worker group: {} from WorkerGroupTaskDispatcherManager error", workerGroup.getName(), e);
```

det101 replied:
Thank you very much for your suggestions! I completely agree with your changes and they do make the code clearer. I will make the changes according to your suggestions.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/queue/TimeBasedTaskExecutionRunnableComparableEntry.java:
```diff
@@ -0,0 +1,75 @@
+
+import org.jetbrains.annotations.NotNull;
+
+public class TimeBasedTaskExecutionRunnableComparableEntry<V> implements Delayed {
```
Since this component is used for `TaskExecutionRunnable` please don't use generic type here or you should use `<V extends ITaskExecutionRunnbale>`

det101 replied:
Thank you very much for your suggestions! I completely agree with your changes and they do make the code clearer. I will make the changes according to your suggestions.

Commit:
optimized code

Commit:
Remove the close logic for WorkerGroupTaskDispatcher. Since this scenario is rare and the logic is simplified for now, it will be left as a TODO.

Commit:
fix unit test

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,150 @@
+        dispatchWorkerMap = new ConcurrentHashMap<>();
+    }
+
+    @PostConstruct
+    public void init() {
+        this.addWorkerGroup(WorkerGroupUtils.getDefaultWorkerGroup());
+    }
```
Seems meaningless
```suggestion
```

det101 replied:
To ensure that the entry point for thread creation in WorkerGroupTaskDispatcher is unique and only triggered through addWorkerGroup, special handling is required for the default worker group since it does not trigger an adding event. Moving the logic to check for the existence of a queue when adding a task could complicate the timing of additions and deletions.  @ruanwenjun

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java:
```diff
@@ -0,0 +1,150 @@
+     */
+    public synchronized boolean addTaskToWorkerGroup(String workerGroup, ITaskExecutionRunnable taskExecutionRunnable,
+                                                     long delayTimeMills) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher != null) {
+            workerGroupTaskDispatcher.addTaskToWorkerGroupQueue(taskExecutionRunnable, delayTimeMills);
+            return true;
+        } else {
+            log.error("workerGroupTaskDispatcher {} not found, will set task {} fail",
+                    workerGroup, taskExecutionRunnable.getTaskInstance().getId());
+        }
+        return false;
```
Directly use computeIfAbsent then you can remove the synchronized

det101 replied:
To ensure that the entry point for thread creation in WorkerGroupTaskDispatcher is unique and only triggered through addWorkerGroup, special handling is required for the default worker group since it does not trigger an adding event. Moving the logic to check for the existence of a queue when adding a task could complicate the timing of additions and deletions.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/MasterServer.java:
```diff
@@ -128,7 +132,9 @@ public void initialized() {
 
         this.masterCoordinator.start();
 
+        this.clusterManager.registerWorkerGroupListener(this.workerGroupTaskDispatcherManager);
```
Meaningless, if we choose to create WorkerGroupDispatcher if not exist in the manager.

det101 replied:
Change the logic to dynamically create the workerGroupTaskDispatcher when adding tasks.

ruanwenjun reviewed (commented):
Basically LGTM

Commit:
Change the logic to dynamically create the workerGroupTaskDispatcher when adding tasks.

Commit:
fix unit test

ruanwenjun commented on dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/runner/queue/TimeBasedTaskExecutionRunnableComparableEntryTest.java:
```diff
@@ -0,0 +1,88 @@
+    @Test
+    public void getDelay_BeforeTriggerTime_ReturnsPositive() {
+        entry = new TimeBasedTaskExecutionRunnableComparableEntry<>(TEST_DELAY_MILLS, testData);
+        sleep(500L);
```
It's better to use `Awaitility`, rather than sleep

det101 replied:
Thank you for your suggestions! I have replaced all instances of sleep with Awaitility. Could you please review my changes? 😊

ruanwenjun reviewed (dismissed):
LGTM, this PR will help to reduce the impact of different worker groups when master dispatch task.

Commit:
optimization unit test

SbloodyS reviewed (commented):
I agree with this implementation, I just have some questions. After reviewing the code, I found that this PR is to put the failed tasks in the delayed queue and wait for a certain period of time before resubmitting, instead of actually dividing them into multiple queues according to the worker group as stated in the title, is that right? @det101

det101 commented:
> I agree with this implementation, I just have some questions. After reviewing the code, I found that this PR is to put the failed tasks in the delayed queue and wait for a certain period of time before resubmitting, instead of actually dividing them into multiple queues according to the worker group as stated in the title, is that right? @det101

If the global queue fails to be distributed to the workergoup queue, the queue will be delayed. If it has been distributed to the workergoup queue, the failure will only be placed in the workergoup's delay queue.
![image](https://github.com/user-attachments/assets/4349f124-3515-42aa-ba6b-d3d514554676)

SbloodyS commented:
> If the global queue fails to be distributed to the workergoup queue, the queue will be delayed. If it has been distributed to the workergoup queue, the failure will only be placed in the workergoup's delay queue.

Okay.

SbloodyS reviewed (approved):
+1
